src/backends/mod.rs

```diff
@@ -35,6 +35,29 @@ pub trait Backend: Clone {
     /// It increments the variable index for look up later.
     fn new_internal_var(&mut self, val: Value<Self>, span: Span) -> CellVar;
 
+    /// basic constraint negation
+    fn enforce_neg_constraint(
+        &mut self,
+        var: &ConstOrCell<Self::Field>,
+        span: Span,
+    ) -> ConstOrCell<Self::Field>;
+
+    /// basic constraint addition
+    fn enforce_add_constraint(
+        &mut self,
+        lhs: &ConstOrCell<Self::Field>,
+        rhs: &ConstOrCell<Self::Field>,
+        span: Span,
+    ) -> ConstOrCell<Self::Field>;
+
+    /// basic constraint multiplication
+    fn enforce_mul_constraint(
+        &mut self,
+        lhs: &ConstOrCell<Self::Field>,
+        rhs: &ConstOrCell<Self::Field>,
+        span: Span,
+    ) -> ConstOrCell<Self::Field>;
```

Comment:
I would call these `neg`, `add`, and `mul` as they might not create a constraint eventually.

Reply:
also as I hint later, I think they should not receive `ConstOrCell` but `CellVar` instead

Reply:
It now uses CellVar at the backend layer